Frameworks/Foundation/NSURLConnection.mm

```diff
@@ -19,168 +20,289 @@
+    [_operationQueue addOperation:[NSBlockOperation blockOperationWithBlock:^void() {
+                         _completionHandler(_response.get(), _data.get(), nil);
+                     }]];
+}
 
-    [state receiveAllDataInMode:mode];
-    [connection unscheduleFromRunLoop:[NSRunLoop currentRunLoop] forMode:mode];
+@end
 
-    id result = state.receivedData;
+// Actual NSURLConnection implementation
+@implementation NSURLConnection {
+@private
+    StrongId<NSObject> _delegate;
 
-    [connection cancel];
+    std::vector<std::pair<StrongId<NSRunLoop>, StrongId<NSString>>> _scheduledRunLoops; // NSRunLoop, NSRunLoopMode
+    // TODO #: Cannot currently schedule on more than one runloop
 
-    if (errorp) {
-        *errorp = state.error;
-    }
+    StrongId<NSOperationQueue> _delegateQueue; // Lazily initialized
 
-    if (responsep) {
-        *responsep = state.response;
-    }
+    StrongId<NSURLResponse> _response;
+    NSURLCacheStoragePolicy _storagePolicy;
+    StrongId<NSURLProtocol> _protocol;
+}
 
-    [connection release];
+/**
+ @Status Interoperable
+*/
++ (BOOL)canHandleRequest:(id)request {
+    return ([NSURLProtocol _URLProtocolClassForRequest:request] != nil) ? YES : NO;
+}
+
+/**
+ @Status Caveat
+ @Notes NSError returned is not detailed
+*/
++ (NSData*)sendSynchronousRequest:(NSURLRequest*)request
+                returningResponse:(NSURLResponse* _Nullable*)response
+                            error:(NSError* _Nullable*)error {
+    if (request) {
+        __block NSData* data = nil;
+        __block bool finished = false;
+        __block NSCondition* condition = [[NSCondition new] autorelease];
+        NSOperationQueue* queue = [[NSOperationQueue new] autorelease];
+
+        [self sendAsynchronousRequest:request
+                                queue:queue
+                    completionHandler:^void(NSURLResponse* innerResponse, NSData* innerData, NSError* innerError) {
+                        if (response) {
+                            *response = innerResponse ? [[innerResponse retain] autorelease] : nil;
+                        }
+
+                        if (error) {
+                            *error = innerError ? [[innerError retain] autorelease] : nil;
+                        }
+
+                        data = innerData ? [[innerData retain] autorelease] : nil;
+
+                        [condition lock];
+                        finished = true;
+                        [condition signal];
+                        [condition unlock];
+                    }];
+
+        [condition lock];
+        while (!finished) {
+            [condition wait];
+        }
+        [condition unlock];
+
+        return data;
+    }
 
-    return result;
+    return nil;
 }
 
 /**
  @Status Interoperable
 */
-+ (id)connectionWithRequest:(id)request delegate:(id)delegate {
++ (NSURLConnection*)connectionWithRequest:(NSURLRequest*)request delegate:(id)delegate {
```

Comment:
nit: this could have `instancetype` as return type (even though the docs don't specify it)